keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/UnhealthyNotificationListener.kt

```diff
@@ -0,0 +1,74 @@
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.events.ResourceHealthEvent
+import com.netflix.spinnaker.keel.events.ResourceNotificationEvent
+import com.netflix.spinnaker.keel.notifications.ClusterViewParams
+import com.netflix.spinnaker.keel.notifications.NotificationMessage
+import com.netflix.spinnaker.keel.notifications.NotificationScope.RESOURCE
+import com.netflix.spinnaker.keel.notifications.Notifier.UNHEALTHY
+import com.netflix.spinnaker.keel.veto.unhealthy.UnsupportedResourceTypeException
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.core.env.Environment
+import org.springframework.stereotype.Component
+
+
+@Component
+class UnhealthyNotificationListener(
+  private val publisher: ApplicationEventPublisher,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String,
+  private val springEnv: Environment
+) {
+
+  private val notificationsEnabled: Boolean
+    get() = springEnv.getProperty("keel.notifications.unhealthy", Boolean::class.java, true)
+
+  @EventListener(ResourceHealthEvent::class)
+  fun onResourceHealthEvent(event: ResourceHealthEvent) {
+    if (notificationsEnabled && !event.healthy){
+      publisher.publishEvent(ResourceNotificationEvent(RESOURCE, event.resource.id, UNHEALTHY, message(event.resource)))
+    }
+  }
+
+  /**
+   *  Assumption: health is only for clusters, and we have specific requirements
+   *  about what the spec looks like in order to construct the notification link
+   *
+   *  Future improvement: add in how long the resource has been unhealthy for
+   */
+  private fun message(resource: Resource<*>): NotificationMessage {
+    val spec = resource.spec
+    if (spec !is Monikered) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be monikered to construct resource links")
+    }
+    if (spec !is Locatable<*>) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be locatable to construct resource links")
+    }
+    val locations = spec.locations
+    if (locations !is AccountAwareLocations) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be have account aware locations to construct resource links")
+    }
+
+    val params = ClusterViewParams(
+      acct = locations.account,
+      q = resource.spec.displayName,
+      stack = spec.moniker.stack ?: "(none)",
+      detail = spec.moniker.detail ?: "(none)"
+    )
+    val resourceUrl = "$spinnakerBaseUrl/#/applications/${resource.application}/clusters?${params.toURL()}"
+
+    return NotificationMessage(
+      subject = "${resource.spec.displayName} is unhealthy",
+      body = "<$resourceUrl|${resource.id}> is unhealthy and " +
+        "Spinnaker does not detect any changes to this resource. " +
```

Comment:
Suggestion:
```suggestion
      body = "Spinnaker has detected that <$resourceUrl|${resource.id}> is unhealthy, but " +
        "there have been no changes to its configuration. " +
```

Reply:
I don't think this is any clearer, and I don't like the use of "but" (a configuration change is not the only reason something could be unhealthy). I'm going to keep it as is for now